### PR TITLE
RemoveProcessor updated to support fieldsToKeep

### PR DESCRIPTION
**Pull request by zembrzuski**
Relates to [#83010](https://github.com/elastic/elasticsearch/issues/83010)

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine ok to test

**Comment by danhermann** [comment]
@zembrzuski, thank you for working on this. In thinking a bit about it, I think the fields to keep should always include the [metadata fields](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestDocument.java#L845) on the document. These are necessary for ingestion to work properly and their removal causes a variety of problems. I'll probably have a few more suggestions after a more thorough review, but I do know this change will be necessary.

**Comment by zembrzuski** [comment]
> @zembrzuski, thank you for working on this. In thinking a bit about it, I think the fields to keep should always include the [metadata fields](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestDocument.java#L845) on the document. These are necessary for ingestion to work properly and their removal causes a variety of problems. I'll probably have a few more suggestions after a more thorough review, but I do know this change will be necessary.

Hi @danhermann
I have just pushed another patch. Now, only fields that are not metadata are eligible for being removed from the document. I also added some unit tests to make sure that every metadata is present in the copy of the document.

Tks so much for reviewing my PR. This is one of my first PRs, and I am very excited about contributing to ElasticSearch.

**Comment by danhermann** [comment]
@elasticmachine update branch

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java [hunk 1]
```diff
@@ -75,8 +102,39 @@ public RemoveProcessor create(
...
+                throw new IllegalArgumentException(
+                    "missing field [processors.remove.keep] or [processors.remove.field]. Please specify one of them."
+                );
+            }
+
+            if (compiledTemplatesToRemove.isEmpty() == false && compiledTemplatesToKeep.isEmpty() == false) {
+                throw new IllegalArgumentException(
+                    "Too many fields specified. Please specify either [processors.remove.keep] or [processors.remove.field]."
+                );
+            }
```
Comment (danhermann) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java:118]: For the sake of consistency, let's use the same exception and message style as other processors with mutually exclusive configuration options. See the [network direction processor](https://github.com/elastic/elasticsearch/blob/master/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/NetworkDirectionProcessor.java#L269) for an example.

File: docs/changelog/83665.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pr: 83665
+summary: "RemoveProcessor updated to support fieldsToKeep"
+area: Infra/Scripting
```
Comment (danhermann) [inline docs/changelog/83665.yaml:3]: Technically, this falls into the "ingest" category:
```suggestion
area: Ingest
```

**Review by danhermann — CHANGES_REQUESTED** [review]
Hi @zembrzuski, thank you for working on this. It looks pretty good. There are just a few changes to be made below and then I think we'll be able to get it merged.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java [hunk 3]
```diff
@@ -27,34 +27,61 @@
...
         this.ignoreMissing = ignoreMissing;
     }
 
-    public List<TemplateScript.Factory> getFields() {
-        return fields;
+    RemoveProcessor(String tag, String description, List<TemplateScript.Factory> fieldsToRemove, boolean ignoreMissing) {
+        super(tag, description);
+        this.fieldsToRemove = new ArrayList<>(fieldsToRemove);
+        this.fieldsToKeep = new ArrayList<>();
+        this.ignoreMissing = ignoreMissing;
```
[the remark this answers is not in the document]
Comment (zembrzuski) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java:51]: ok

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 4]
```diff
@@ -859,6 +879,10 @@ public String toString() {
             this.fieldName = fieldName;
         }
 
+        public static boolean isMetadata(String field) {
+            return Arrays.stream(Metadata.values()).anyMatch(e -> e.fieldName.equals(field));
+        }
```
[the remark this answers is not in the document]
Comment (zembrzuski) [reply server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:884]: ok

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java [hunk 5]
```diff
@@ -27,34 +27,61 @@
...
 
     @Override
     public IngestDocument execute(IngestDocument document) {
+        if (fieldsToKeep.isEmpty() == false) {
+            IngestDocument.getAllFields(document.getSourceAndMetadata())
+                .stream()
+                .filter(documentField -> IngestDocument.Metadata.isMetadata(documentField) == false)
+                .filter(documentField -> shouldKeep(documentField, fieldsToKeep, document) == false)
+                .forEach(documentField -> removeWhenPresent(document, documentField));
+        }
```
[the remark this answers is not in the document]
Comment (zembrzuski) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java:62]: ok

File: docs/changelog/83665.yaml (hunk 2 above)
Comment (zembrzuski) [reply docs/changelog/83665.yaml:3]: ok

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java (hunk 1 above)
Comment (zembrzuski) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemoveProcessor.java:118]: ok

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 6]
```diff
@@ -786,6 +787,25 @@ public static Object deepCopy(Object value) {
...
+    @SuppressWarnings("unchecked")
+    private static Set<String> getAllFields(Map<String, Object> input, String prefix) {
+        Set<String> allFields = Sets.newHashSet();
+
+        input.forEach((k, v) -> {
+            allFields.add(prefix + k);
+
+            if (v instanceof Map) {
+                allFields.addAll(getAllFields((Map) v, prefix + k + "."));
+            }
```
[the remark this answers is not in the document]
Comment (zembrzuski) [reply server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:803]: ok

**Comment by zembrzuski** [comment]
Hi @danhermann 
Thanks for reviewing this PR! I've just pushed the code changes asked in the code review.

**Comment by danhermann** [comment]
@zembrzuski, I think you just need to update the error message in the RemoveProcessorFactory tests so they pass.

**Comment by zembrzuski** [comment]
> @zembrzuski, I think you just need to update the error message in the RemoveProcessorFactory tests so they pass.

Ohh, my bad. I've just fixed these tests.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/rest-compatibility

**Comment by danhermann** [comment]
@zembrzuski, everything here looks good. One minor change needed and then we can get this merged:

On the following two pairs of lines, `processors.remove.keep` should just be `keep` and `processors.remove.field` should just be `field`:
https://github.com/elastic/elasticsearch/pull/83665/files#diff-e44059a95b4557dd2e3c310c64cca96ce1c0dc1f479cfa40c2c32634a91c8512R116-R117 

https://github.com/elastic/elasticsearch/pull/83665/files#diff-e44059a95b4557dd2e3c310c64cca96ce1c0dc1f479cfa40c2c32634a91c8512R125-R126

That will change the error messages a bit, so a couple of the unit tests will likely need to be updated accordingly.

**Comment by zembrzuski** [comment]
> 

Hi @danhermann 
I've just updated the messages and the unit tests! Tks!

**Comment by danhermann** [comment]
> Hi @danhermann I've just updated the messages and the unit tests! Tks!

Thanks, @zembrzuski. It looks like the formatting check is complaining about a couple lines now. You should be able to fix that by running `./gradlew :modules:ingest-common:spotlessApply`.

**Comment by zembrzuski** [comment]
> > Hi @danhermann I've just updated the messages and the unit tests! Tks!
> 
> Thanks, @zembrzuski. It looks like the formatting check is complaining about a couple lines now. You should be able to fix that by running `./gradlew :modules:ingest-common:spotlessApply`.

Ohh, I'm sorry. I've read about checkstyle in the contribution guideline but I forgot to check it this time.

**Comment by ddolcimascolo** [comment]
Hey guys, so this landed in 8.2 right? Is it possible that the documentation was forgotten? I can't find anything related to the new `keep` configuration option in https://www.elastic.co/guide/en/elasticsearch/reference/current/remove-processor.html

I'll give it a try anyway :)

**Pull request merged**